File: app/src/main/java/com/thewhitewings/pouch/ui/NoteViewModel.java
package com.thewhitewings.pouch.ui;

import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;

import com.thewhitewings.pouch.data.Note;
import com.thewhitewings.pouch.data.NotesRepository;
import com.thewhitewings.pouch.utils.Constants;

/**
 * ViewModel to interact with the {@link NotesRepository}'s data source and the Note screen.
 */
public class NoteViewModel extends ViewModel {

    private static final String TAG = "NoteViewModel";
    private final NotesRepository notesRepository;

    // the note that is opened for updating purpose
    private Note oldNote;

    // holds the state of the currently opened note
    private final MutableLiveData<Note> noteLiveData = new MutableLiveData<>();

    /**
     * Constructor for {@link NoteViewModel}.
     *
     * @param repository the repository to interact with the data source
     */
    public NoteViewModel(NotesRepository repository) {
        this.notesRepository = repository;
    }

    /**
     * Get the state of note LiveData.
     *
     * @return the note LiveData
     */
    public LiveData<Note> getNoteLiveData() {
        return noteLiveData;
    }

    /**
     * Update the state of note LiveData with the provided note.
     *
     * @param note the note with updated data
     */
    public void updateNoteLiveData(Note note) {
        noteLiveData.setValue(note);
    }

    /**
     * Initialize the note LiveData and {@link #oldNote} with the provided arguments.
     * It is needed for the first initialization when a note is opened for updating.
     *
     * @param args the arguments bundle passed from the activity containing the opened note data
     */
    public void initializeNote(Bundle args) {
        if (args == null) // creating new note
            return;

        if (noteLiveData.getValue() != null) // updating existing note - after configuration change
            return;

        // updating existing note - first initialization
        int id = args.getInt(Constants.COLUMN_ID);
        String noteTitle = args.getString(Constants.COLUMN_NOTE_TITLE);
        String noteBody = args.getString(Constants.COLUMN_NOTE_BODY);
        String timestamp = args.getString(Constants.COLUMN_TIMESTAMP);

        oldNote = new Note(id, noteTitle, noteBody, timestamp);
        updateNoteLiveData(oldNote);
    }


    /**
     * Create or update the note based on the provided data.
     *
     * @param newNoteTitle the new title of the note
     * @param newNoteBody  the new body of the note
     */
    public void createOrUpdateNote(String newNoteTitle, String newNoteBody) {
        if (oldNote == null && (!newNoteTitle.isEmpty() || !newNoteBody.isEmpty()))
            createNote(newNoteTitle, newNoteBody);
        else if (oldNote != null &&
                (!oldNote.getNoteBody().equals(newNoteBody) || !oldNote.getNoteTitle().equals(newNoteTitle)))
            updateNote(newNoteTitle, newNoteBody);
    }

    /**
     * Create a new note with the provided title and body.
     *
     * @param noteTitle the title of the new note
     * @param noteBody  the body of the new note
     */
    private void createNote(String noteTitle, String noteBody) {
        notesRepository.createNote(noteTitle, noteBody);
    }

    /**
     * Update the existing note with the provided title and body.
     *
     * @param noteTitle the new title of the note
     * @param noteBody  the new body of the note
     */
    private void updateNote(String noteTitle, String noteBody) {
        if (noteLiveData.getValue() != null)
            notesRepository.updateNote(noteTitle, noteBody, noteLiveData.getValue().getId());
    }

    /**
     * Delete the currently opened note.
     */
    public void deleteNote() {
        if (noteLiveData.getValue() != null)
            notesRepository.deleteNote(noteLiveData.getValue());
    }


    /**
     * Factory class for creating instances of {@link NoteViewModel}.
     */
    public static class NoteViewModelFactory implements ViewModelProvider.Factory {
        private final NotesRepository repository;

        public NoteViewModelFactory(NotesRepository repository) {
            this.repository = repository;
        }

        @NonNull
        @Override
        public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
            if (modelClass.isAssignableFrom(NoteViewModel.class)) {
                return (T) new NoteViewModel(repository);
            }
            throw new IllegalArgumentException("Unknown ViewModel class");
        }
    }
}

File: app/src/main/java/com/thewhitewings/pouch/ui/NoteViewModel.kt
package com.thewhitewings.pouch.ui

import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.ViewModelProvider.AndroidViewModelFactory.Companion.APPLICATION_KEY
import androidx.lifecycle.createSavedStateHandle
import androidx.lifecycle.viewModelScope
import androidx.lifecycle.viewmodel.initializer
import androidx.lifecycle.viewmodel.viewModelFactory
import com.thewhitewings.pouch.PouchApplication
import com.thewhitewings.pouch.data.Note
import com.thewhitewings.pouch.data.NotesRepository
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

private const val TAG = "NoteViewModel"

/**
 * ViewModel to interact with the [NotesRepository]'s data source and the Note screen.
 */
class NoteViewModel(
    private val savedStateHandle: SavedStateHandle,
    private val notesRepository: NotesRepository,
    private val dispatcher: CoroutineDispatcher = Dispatchers.IO
) : ViewModel() {

    // the note that is opened for updating purpose
    private var oldNote: Note? = null

    // Holds current NoteUiState
    private val _noteUiState = MutableStateFlow(NoteUiState())
    val noteUiState = _noteUiState.asStateFlow()

    init {
        initializeNote()
    }

    /**
     * Initialize the note with the given ID when opening an existing note.
     */
    private fun initializeNote() {
        val noteId: Int = savedStateHandle[NoteDestination.noteIdArg]
            ?: return

        viewModelScope.launch(dispatcher) {
            notesRepository.getNoteById(noteId)
                .collect { note ->
                    if (note != null) {
                        _noteUiState.value = _noteUiState.value.copy(
                            note = note
                        )
                        oldNote = note
                    }
                }
        }
    }

    /**
     * Updates the note title of the current note's state
     */
    fun updateNoteTitle(title: String) {
        _noteUiState.update {
            it.copy(
                note = it.note.copy(noteTitle = title)
            )
        }
    }

    /**
     * Updates the note body of the current note's state
     */
    fun updateNoteBody(body: String) {
        _noteUiState.update {
            it.copy(
                note = it.note.copy(noteBody = body)
            )
        }
    }

    /**
     * Create a new note or update the currently opened note.
     *
     * To create a new note, the note title or note body must not be empty.
     *
     * To update a note, the note title and note body must not be the same as the old note.
     * Otherwise, the note will not be updated, i.e., its timestamp will stay the same.
     */
    fun createOrUpdateNote() {
        with(_noteUiState.value.note) {
            if (oldNote == null) {
                if (noteTitle.isNotEmpty() || noteBody.isNotEmpty())
                    createNote()
            } else if (!oldNote?.noteBody.equals(noteBody) ||
                !oldNote?.noteTitle.equals(noteTitle)
            ) {
                updateNote()
            }
        }
    }

    /**
     * Create a new note.
     */
    private fun createNote() {
        viewModelScope.launch(dispatcher) {
            notesRepository.createNote(_noteUiState.value.note)
        }
    }

    /**
     * Update an existing note.
     */
    private fun updateNote() {
        viewModelScope.launch(dispatcher) {
            notesRepository.updateNote(_noteUiState.value.note)
        }
    }

    /**
     * Deletes a note.
     */
    fun deleteNote() {
        viewModelScope.launch(dispatcher) {
            notesRepository.deleteNote(_noteUiState.value.note)
        }
    }


    /**
     * UI state for the Note screen
     */
    data class NoteUiState(
        val note: Note = Note(timestamp = "")
    )


    companion object {

        /**
         * Factory for [NoteViewModel] that takes [NotesRepository] as a dependency
         */
        val Factory: ViewModelProvider.Factory = viewModelFactory {
            initializer {
                val application = (this[APPLICATION_KEY] as PouchApplication)
                val notesRepository = application.notesRepository
                NoteViewModel(
                    savedStateHandle = this.createSavedStateHandle(),
                    notesRepository = notesRepository
                )
            }
        }
    }
}
